lib/utils.pm

```diff
@@ -1658,6 +1659,44 @@ sub script_output_retry {
     die("Waiting for Godot: $cmd") if $die;
 }
 
+
+=head2 validate_script_output_retry
+
+ validate_script_output_retry($cmd, $check, [retry => $retry], [delay => $delay], [timeout => $timeout]);
+
+Repeat command until validate_script_output succeeds or die. Return the output of the command on success.
+
+C<$retry> refers to the number of retries and defaults to C<10>.
+
+C<$delay> is the time between retries and defaults to C<30>.
+
+If the command doesn't succeed after C<$retry> retries,
+this function will die.
+
+Example:
+
+ validate_script_output_retry('ping -c1 -W1 machine', m/1 packets transmitted/, retry => 5, delay => 60);
+
+=cut
+sub validate_script_output_retry {
+    my ($cmd, $check, %args) = @_;
+    my $retry = $args{retry} // 10;
+    my $delay = $args{delay} // 30;
+    my $timeout = $args{timeout} // 90;
+
+    for (1 .. $retry) {
+        eval { validate_script_output($cmd, $check, timeout => $timeout); };
+        if ($@) {
```

Comment:
I think it is crucial here to figure out WHY exactly it failed . We don't want to retry when script simply failed because validation failed . ONLY good reason to fail is due to timeout

Reply:
Very good point... any ideas how?

Reply:
I also think in case of timeout, simply retry will cause a console mess, as you are not able to judge if you got the prompt again.
in `wicked::wlan::retry` we do it this way: https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/lib/wicked/wlan.pm#L98
With `wicked::wlan:retry()` can can actually call every testapi console commands like: https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/lib/wicked/wlan.pm#L282 so you do not need to write a wrapper for all of these commands.


Or you go this `timeout $cmd` approach from other `retry()` functions. https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/lib/utils.pm#L1604

But in general, I think it is better to check some preconditions and then the real command should work. Simply retry might actually hide some errors.

Reply:
ok, I'll go `timeout $cmd` way. Thanks! What I'm not sure is how to differenciate that the eval failed due to command timeout (to retry) or due to output not validating (to avoid retrying).

Reply:
I might be mistaken, but `$@` here would have the exception that's thrown by the `croak` in the `wait_serial`  that has the timeout in `distribution::script_output`:

```perl
use Carp;
use feature 'say';
eval {
# warn user (from perspective of caller)
carp "string trimmed to 80 chars";
 
# die of errors (from perspective of caller)
croak "We're outta here!";
};

if ($@){
	say $@;
	say "oops" if $@ =~ /outta/;
}
```

Reply:
ACK. But it isn't nice, cause your code will fail, once someone change `validated_script_output()`

```suggestion
        return 0 unless defined($@);
        die($@) if ($@ ne 'validate_script_output');
        
        record(...); sleep xxx;
```

Reply:
For some reason, `$@` is always defined, whatever happens inside the eval (succeeds or fails).

Reply:
@jlausuch it's meant to: [form perldoc](https://perldoc.perl.org/perlvar#%24%40)

> If no error occurs, eval sets $@ to the empty string.

Reply:
thx @foursixnine wasn't aware of that detail !